### PR TITLE
[Jobs] [Doc] Add Jobs openapi spec and REST API documentation

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
- Adds an OpenAPI spec for the Jobs REST API
- Adds an API doc page using Redoc.

The spec was generated via a combination of pydantic automatic schema generation, github copilot, and manual entry, but this won't be maintainable in the long term. When we move the dashboard server to FastAPI, this spec can be automatically generated.

Link to the doc page built from this PR: https://ray--30417.org.readthedocs.build/en/30417/cluster/running-applications/job-submission/rest.html

Screen Shot 2022-11-17 at 8 52 02 AM

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (architkulkarni) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:4]: The `/api/version` returns `"4"`, not `"4.0.0"`.  Do these need to be kept in sync?

File: doc/requirements-doc.txt
Comment (architkulkarni) [inline doc/requirements-doc.txt:62]: I don't have a particular reason to pin the version here, just following what's been done for other sphinx extensions

**Comment by edoakes** [comment]
@architkulkarni re: the process to generate the documentation, at a minimum we need to document how to do this in case others make changes in the future. Could you add a short README somewhere in the code?

**Comment by edoakes** [comment]
This is fantastic though!

**Comment by architkulkarni** [comment]
@edoakes Good point, I added a note to `job_head.py`

File: dashboard/modules/job/job_head.py
Comment (edoakes) [inline dashboard/modules/job/job_head.py:151]: I wouldn't know how to do this from the description alone, can you include instructions?

Comment (architkulkarni) [reply dashboard/modules/job/job_head.py]: @edoakes updated, let me know if any parts are unclear

**Review by DmitriGekhtman — APPROVED** [review]
stamp

**Review by edoakes — APPROVED** [review]
Suggest to get a review from @alanwguo before merging

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (alanwguo) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:1]: is this file manually written? Did we consider switching the jobs api to use fast-api which has the ability to automatically generate openapi spec from the route definitions?

@rkooo567 and I have been chatting about doing a fast-api migration in the next 6 months so that we can have openapi specs for all of our APIs. If we are to do that, there's a chance that this spec will be slightly modified because it's generated instead of manually written. Will that be disruptive to users?

Comment (architkulkarni) [reply doc/source/cluster/running-applications/job-submission/openapi.yml:1]: @alanwguo Yeah, migrating Jobs to FastAPI makes sense as the long term plan, we just didn't have time to do it for Ray 2.2.  I think by marking the OpenAPI section of the doc as "Beta" as it currently is marked this PR, it should be okay if there are slight modifications in the future. The underlying API is stable and once we migrate to OpenAPI we can mark the OpenAPI doc as "stable" as well.

File: dashboard/modules/job/job_head.py
Comment (dmatrix) [inline dashboard/modules/job/job_head.py:155]: The sentence is long. I would break it. 

"update it manually." Also, bump the version number in the yaml file and in this class's `get_version`.

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (dmatrix) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:16]: Should we use possessive here? Ray Jobs' API? If so then change in all description below.
I guess this would be tantamount to asking "Get Python's version" or "Get Python version." 

I think either would work. But a grammar pedantic would suggest otherwise.

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (dmatrix) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:28]: Which one reads declaratively:
"The version of the Ray Jobs API running on the server." or
"Ray Jobs' API version running on the server."

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (dmatrix) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:286]: Ending with a proposition could read awkward in some cases. Would this be simpler: "Tail job ID's logs."

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (dmatrix) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:330]: launched in Ray.

Comment (dmatrix) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:330]: This can --> This id can be used to ...

File: doc/source/cluster/running-applications/job-submission/openapi.yml
Comment (dmatrix) [inline doc/source/cluster/running-applications/job-submission/openapi.yml:336]: The antecedent is ambiguous here. Does it refer to id or job. It normally refers to the last antecedent, which is job, and that is incorrect. 

Suggest use "This id can be used to fetch ...

**Review by dmatrix — APPROVED** [review]
minor nits. You can incorporate them if they make sense or add clarity. LGTM

**Comment by architkulkarni** [comment]
Thanks @dmatrix, I will incorporate these in a follow-up PR!

**Pull request merged**
